Multi AZ storage for OSD (#14917)
openshift/openshift-docs: architecture/additional_concepts/storage.adoc

        claimName: myclaim
----

ifdef::openshift-dedicated[]
[[osd-multi-az-pv]]

> == OSD Multi-AZ Cluster PV Usage Examples

Reviewer: ```suggestion
== Multi-AZ persistent volume claims
```

> A walk through of common use cases for persistent volumes on a Multi-AZ

Reviewer: ```suggestion
A walkthrough of common use cases for persistent volumes on a Multi-AZ
```

OpenShift Dedicated cluster.
[[osd-multi-az-pv-config]]

> === Configuration Details

Reviewer: ```suggestion
=== Configuration details
```

   displayName: Storage class name
   name: 'STORAGE_CLASS_NAME'
   required: true
----

Now that you have an updated template, you can use it from the CLI or web console.
From the CLI, you can pass in the required parameters to “oc process” and
 create them:

----
oc process -f mongodb-export.yml \
STORAGE_CLASS_NAME=gp2-encrypted-us-east-2b | oc create -f
----

Or you can upload your template via the web console and fill out the parameters
  of the template interactively. To do that, click the “Add to Project” dropdown
in the application console and select the option “Import YAML / JSON”:

image::add-to-project.png[Import yaml/json to project]

From here, you can either upload your template file or copy/paste it into the
provided text field and then click create. From here you can opt to process
and/or save your template file to the selected namespace for future use.
If you process the template, you’ll be taken to a page to fill out the available
 template parameters, including the desired storage class. After populating the parameters,
you can create the mongodb service and confirm that your PVC has been created in
 the correct AZ.

[[osd-multi-az-pvc-best-practices]]

> === Best Practices

Reviewer: ```suggestion
=== Best practices
```

        claimName: myclaim
----

ifdef::openshift-dedicated[]

[[osd-multi-az-pv]]
== Multi-AZ persistent volume claims

A walkthrough of common use cases for persistent volumes on a Multi-AZ
{product-title} cluster.

[[osd-multi-az-pv-config]]
=== Configuration details

> A multi-AZ {product-title} cluster will come configured with one storage class per

Reviewer 2: s/will come/comes
s/will be set/is set

classes will be set as the default. For example, a cluster provisioned in
us-east-2 would have three storage classes that are named to indicate their
availability zone.

.StorageClass availability zone
[cols="3,2,1",options="header"]
|===
|StorageClass Name |Provisioner |Age
|gp2-encrypted-us-east-2a (default)
|kubernetes.io/aws-ebs
|3d
|gp2-encrypted-us-east-2b
|kubernetes.io/aws-ebs
|3d
|gp2-encrypted-us-east-2c
|kubernetes.io/aws-ebs
|3d
|===

[[osd-multi-az-pv-usecases]]
=== Use cases

[[osd-multi-az-pv-usecases-pvc-storageclass-cli]]
==== Creating a PersistentVolumeClaim with a specific storage class (CLI)

Creating a PersistentVolumeClaim (PVC) via the CLI requires a manifest file that
provides the name and details of the PersistentVolume (PV) being requested.
To have the PVC use a different storage class, simply indicate that in your
manifest.

.PVC example
[source,yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mypvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
  requests:
    storage: 1Gi
  storageClassName: gp2-encrypted-us-east-2b <1>
----
<1> Specify your storage class to override the default.

> Passing this manifest to `oc create -f <manifest filename>` will create a PV in

Reviewer 2: s/will create/creates

the AZ of the storage class provided. You can confirm this by checking the
output of `oc get pvc`.

[[osd-multi-az-pv-usecases-pvc-storageclass-ui]]
==== Creating a PersistentVolumeClaim with a specific storage class (UI)

> The web console UI makes choosing a storage class for your PVC easy.

Reviewer 2: Remove the first sentence, as we should not indicate what is easy for the user. Consider replacing with:

    To choose a storage class for your PVC using the web console, navigate to the *Storage* dashboard and click 
*Create Storage*.

In the application console, navigate to the Storage dashboard and click the
"Create Storage" button. The first option on the "Create Storage" page is a

> dropdown list of available storage classes to select from.

Reviewer 2: drop-down list of available storage classes.

image::storage-class.png[Create storage with UI,title="Storage Class dropdown"]

> The dropdown menu shows the name, zone, and access mode for the available

Reviewer 2: drop-down menu

storage classes. You can select your desired storage class, provide a name and

> size for the PVC, and click "Create".

Reviewer 2: *Create*.

[[osd-multi-az-pv-usecases-pvc-non-default-storageclass-cli]]
==== Using non-default storage classes with OpenShift templates (CLI only)

> OpenShift ships with a number of templates that can be used to quickly deploy

Reviewer 2: s/OpenShift/{product-title}

popular applications and frameworks. These are listed in the catalog when using

> the web console and are viewable from CLI by viewing the output of.

Reviewer 2: output of:

----
$ oc get templates -n openshift
----

Some of these templates include PVC definitions to provide persistent storage,

> but they do not natively support selection of a non-default storage class

Reviewer 2: class. However, there are plans to support this in the future.

(however there is an RFE filed to support this in the future).

> To make use of a template with a specific storage class, you'll need to copy the

Reviewer 2: To use a template with a specific storage class, you need

> template to disk and edit it prior to using it. For example, if you wanted to

Reviewer 2: s/wanted/want

use the MongoDB template with a PVC in us-east-2b, you would first copy the

> template to a file.

Reviewer 2: us-east-2b: 
    
    .Copy the template to a file:
    +
    ----
    $ oc get template -n openshift mongodb-persistent -o yaml --export > mongodb-persistent.yml
    ----
   
    .  Look through . . .

----

Then, look through the file for any PersistentVolumeClaim definitions and edit
them to include an option to specify the storage class. The MongoDB template has
one PVC defined. The line highlighted below shows what needs to be added to allow
specification of a non-default storage class.

.PVC Non-default StorageClass example
[source,yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ${DATABASE_SERVICE_NAME}
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: ${VOLUME_CAPACITY}
  storageClassName: ${STORAGE_CLASS_NAME}
----

Then, at the bottom of the template file, add the new parameter
`STORAGE_CLASS_NAME` to the parameter list.

.PVC Non-default StorageClass example contd
[source,yaml]
----
description: Storage class to use for persistent storage
displayName: Storage class name
name: STORAGE_CLASS_NAME
required: true
----

Now that you have an updated template, you can use it from the CLI or web console.
From the CLI, you can pass in the required parameters to `oc process` and
create them.

----
$ oc process -f mongodb-export.yml \
STORAGE_CLASS_NAME=gp2-encrypted-us-east-2b | oc create -f
----

> Or you can upload your template via the web console and fill out the parameters

Reviewer 2: s/Or/Alternatively,

> of the template interactively. To do that, click the "Add to Project" dropdown

Reviewer 2: To do so, select *Import YAML / JSON* from the *Add to Project* drop-down menu in the application console.

in the application console and select the option "Import YAML / JSON".

image::add-to-project.png[Import yaml/json to project,title="Import yaml/json to project"]

> From here, you can either upload your template file or copy/paste it into the

Reviewer 2: s/From here/Now,

Reviewer 2: copy and paste it

> provided text field and then click create. From here you can opt to process

Reviewer 2: click *Create*

Reviewer 2: s/From here/Now,
s/you'll be taken/you are directed to a page

and/or save your template file to the selected namespace for future use.
If you process the template, you'll be taken to a page to fill out the available
template parameters, including the desired storage class. After populating the parameters,
you can create the MongoDB service and confirm that your PVC has been created in
the correct AZ.

[[osd-multi-az-pvc-best-practices]]
=== Best practices

It is very important to deliberately select a storage class when provisioning

> PVCs in a multi-AZ OSD cluster. If every PVC is always provisioned by the

Reviewer 2: s/OSD/{product-title}

default storage class, this can lead to problems in the future.

One of these problems is an uneven distribution of workload. If you are
primarily running workloads that make use of persistent storage, and you always
provision that storage using the default storage class, all of your pods will be
scheduled to nodes in that AZ. This can lead to a situation where the nodes in one
AZ are running out of capacity while the rest of the cluster is underutilized.

> Once you are in this state, there isn't an easy method to redistribute the workloads.

Reviewer 2: s/isn't/is not

Any redistribution will require backing up and restoring the contents of PVs manually.

> The other problem with winding up in this state is that it effectively undermines

Reviewer 2: s/with winding up in this state/with this state
s/goes down/fails
s/go down/be lost
